### PR TITLE
AEA-1631 libp2p acn part refactoring

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

libp2p acn part refactoring

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes and CI passes too
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: libs/go/libp2p_node/acn/utils.go [hunk 1]
```diff
@@ -181,3 +200,260 @@ type Pipe interface {
...
+	}
+	buf, err := proto.Marshal(response)
+	if err != nil {
+		return err
+	}
+	err = pipe.Write(buf)
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (DavidMinarsch) [inline libs/go/libp2p_node/acn/utils.go:365]: why not simple `return err`?

File: libs/go/libp2p_node/acn/utils.go [hunk 2]
```diff
@@ -181,3 +200,260 @@ type Pipe interface {
...
+	messageType, envelope, _, acnErr := DecodeAcnMessage(buf)
+
+	if acnErr != nil {
+		err = SendAcnError(
+			pipe,
+			acnErr.Error(),
+			acnErr.ErrorCode,
+		)
+		ignore(err)
+		return nil, acnErr.Err
```
Comment (DavidMinarsch) [inline libs/go/libp2p_node/acn/utils.go:453]: no error message?

File: libs/go/libp2p_node/dht/dhtclient/dhtclient.go [hunk 3]
```diff
@@ -463,89 +463,35 @@ func (dhtClient *DHTClient) RouteEnvelope(envel *aea.Envelope) error {
...
-			Msg("while receiving response acn message")
+			Msg("while sending LookupRequest message")
 		errReset := stream.Reset()
 		ignore(errReset)
 		return err
 	}
 
-	response := &acn.AcnMessage{}
-	err = proto.Unmarshal(buf, response)
+	record, err := acn.ReadLookupResponse(streamPipe)
```
Comment (DavidMinarsch) [inline libs/go/libp2p_node/dht/dhtclient/dhtclient.go:479]: Given they are always used together, why not wrap it into a method in utils?

File: libs/go/libp2p_node/dht/dhtclient/dhtclient.go [hunk 4]
```diff
@@ -881,51 +713,6 @@ func (dhtClient *DHTClient) handleAeaAddressStream(stream network.Stream) {
...
-	default:
-		err = errors.New("Unexpected payload, expected LookupRequest")
-		status := &acn.StatusBody{Code: acn.ERROR_UNEXPECTED_PAYLOAD, Msgs: []string{err.Error()}}
-		statusPerformative := &acn.StatusPerformative{Body: status}
-		response := &acn.AcnMessage{
-			Performative: &acn.Status{Status: statusPerformative},
-		}
-		buf, err = proto.Marshal(response)
-		ignore(err)
-		err = utils.WriteBytes(stream, buf)
```
Comment (DavidMinarsch) [inline libs/go/libp2p_node/dht/dhtclient/dhtclient.go:920]: where is this done now? in utils I guess..worth checking

**Review by DavidMinarsch — COMMENTED** [review]
LGTM, there are a few comments, also this test (tests/test_packages/test_connections/test_p2p_libp2p/test_fault_tolerance.py::TestLibp2pConnectionRelayNodeRestartIncomingEnvelopes::test_envelope_routed_from_client_after_relay_restart) fails across the macOS runs

File: libs/go/libp2p_node/acn/utils.go (hunk 1 above)
Comment (solarw) [reply libs/go/libp2p_node/acn/utils.go:365]: good point

File: libs/go/libp2p_node/acn/utils.go [hunk 5]
```diff
@@ -162,13 +169,25 @@ func SendAcnError(pipe Pipe, error_msg string, err_codes ...Status_ErrCode) erro
...
 		Performative: &AeaEnvelope{AeaEnvelope: performative},
 	}
+
 	buf, err := proto.Marshal(msg)
-	return err, buf
+	if err != nil {
+		logger.Error().
+			Str("err", err.Error()).
+			Msgf("while serializing envelope bytes: %s", envelope_bytes)
+	}
```
[the remark this answers is not in the document]
Comment (solarw) [reply libs/go/libp2p_node/acn/utils.go:189]: just extra logging, should not be a problem

File: libs/go/libp2p_node/acn/utils.go (hunk 2 above)
Comment (solarw) [reply libs/go/libp2p_node/acn/utils.go:453]: this is ok, errors are logged in function called above, so we just propagate error raised on decode acn message

File: libs/go/libp2p_node/acn/utils.go [hunk 6]
```diff
@@ -181,3 +200,260 @@ type Pipe interface {
...
+	buf, err := proto.Marshal(msg)
+	if err != nil {
+		return err
+	}
+	err = pipe.Write([]byte(buf))
+	if err != nil {
+		return err
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (solarw) [reply libs/go/libp2p_node/acn/utils.go:314]: done

File: libs/go/libp2p_node/aea/api.go [hunk 7]
```diff
@@ -430,48 +429,45 @@ func (aea *AeaApi) stop() {
...
-		return err
-	}
-
-	status, err := acn.WaitForStatus(aea.acn_status_chan, AcnStatusTimeout)
+	err = acn.SendEnvelopeMessageAndWaitForStatus(
+		pipe,
+		envelope_bytes,
+		acn_status_chan,
+		acnStatusTimeout,
+	)
```
[the remark this answers is not in the document]
Comment (solarw) [reply libs/go/libp2p_node/aea/api.go:462]: cause dependencies, acn know nothing about aea.Envelope actually.
acn knows about envelope bytes, but it's decoded to aea.Envelope.
just need to find out how to move dependencies properly.

File: libs/go/libp2p_node/dht/dhtclient/dhtclient.go (hunk 3 above)
Comment (solarw) [reply libs/go/libp2p_node/dht/dhtclient/dhtclient.go:479]: nice point

File: libs/go/libp2p_node/dht/dhtclient/dhtclient.go (hunk 4 above)
Comment (solarw) [reply libs/go/libp2p_node/dht/dhtclient/dhtclient.go:920]: it's done in utils with helper methods readrequest/sendresponse functions

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**